### PR TITLE
[optional] support loading and saving models with protobuf

### PR DESCRIPTION
**Pull request by wxchan**
related to #372 

- [x] solve conflicts
- [x] test windows support: mingw can work, vs not tested yet

### CODE REVIEW

**Comment by guolinke** [comment]
@wxchan 
the license of protobuf? and is it cross platform ?

**Comment by wxchan** [comment]
@guolinke 

LICENSE: https://github.com/google/protobuf/blob/master/LICENSE

I add a test and it passes on both linux and osx, I am not sure about windows.

test on osx: https://travis-ci.org/wxchan/LightGBM/jobs/274043104, it takes too long.

**Comment by guolinke** [comment]
@wxchan what is the compression ratio of using protobuf, compared with raw text format ? 

I think using protobuf is too heavy, can we just using a simple binary format (like the dataset bin file) ?

**Comment by wxchan** [comment]
@guolinke I didn't test on big model yet. My computer takes too long to load big datasets. The good thing of protobuf is, it's readable through .proto file. We can easily know what features are saved.

**Comment by wxchan** [comment]
@guolinke I made a test on Higgs dataset

text ->proto
after 500 rounds
save: 0.612300s -> 0.023627s
load: 0.471999s -> 0.024712s
size: 13M -> 5.9M

File: .travis.yml [hunk 1]
```diff
@@ -36,6 +37,16 @@ matrix:
       env: TASK=gpu METHOD=pip
     - os: osx
       env: TASK=pylint
+    - os: osx
+      env: TASK=if-else
+    - os: osx
+      env: TASK=proto
+    - os: osx
+      env: TASK=regular PYTHON_VERSION=2.7
```
Comment (henry0312) [inline .travis.yml]: please see https://github.com/Microsoft/LightGBM/pull/904#issuecomment-328732939

Comment (wxchan) [reply .travis.yml]: @henry0312 I saw. Do you have an agreement yet? I can add it later, I think it's a waste of resource to run those python tests I didn't alter multiple time now.

Comment (henry0312) [reply .travis.yml]: > Do you have an agreement yet?

No, I don't know if @guolinke aggrees with it yet, but I think we should run tests with Python 2.7.x and Python 3.6.x at least.

> I think it's a waste of resource to run those python tests I didn't alter multiple time now.

I got it!
Yeah, Travis always faces with lacks of instances of OSX.
We have no choice but to reduce test cases for OSX 😢

**Comment by wxchan** [comment]
one similar structure: https://github.com/dmlc/treelite/blob/master/src/tree.proto

**Comment by guolinke** [comment]
@wxchan any updates of this ? 
this could be merged. but we still need a cross platform solution.

**Comment by guolinke** [comment]
@wxchan it seems that the treelite is like our if-else solution ?

**Comment by wxchan** [comment]
@hcho3 
have you used this proto https://github.com/dmlc/treelite/blob/master/src/tree.proto to compress LightGBM model? what's the compression ratio?

**Comment by hcho3** [comment]
@wxchan No, in treelite, protobuf is only for input not for output. (Protobuf is one of four methods to input a model into treelite compiler.) So LightGBM model isn't being translated into protobuf.

**Comment by Tony-Y** [comment]
I found useful information: https://github.com/thekvs/cpp-serializers
I think [cereal](http://uscilab.github.io/cereal/index.html) is better than protobuf.

**Comment by wxchan** [comment]
@guolinke I already finished this thread.

**Comment by guolinke** [comment]
@wxchan this PR will break the build of VS. I will revert it and merge back after fixing this.

**Comment by wxchan** [comment]
I think I forgot changing some path of file in windows/LightGBM.vcxproj after moving it, is this the reason?

**Comment by guolinke** [comment]
@wxchan it cannot find some ".h" and ".cpp" files, seems are some the auto-generated protobuf files.

**Comment by wxchan** [comment]
@guolinke maybe this one https://github.com/wxchan/LightGBM/blob/1e6091c01f75cbbc86b55b104ae3231d302a0a9f/src/proto/not_implemented/model.pb.h?

**Comment by guolinke** [comment]
@wxchan 
yeah, it says cannot find model.pb.h ... .
Maybe forget to add it to these include path ? https://github.com/Microsoft/LightGBM/blob/master/windows/LightGBM.vcxproj#L63

**Comment by wxchan** [comment]
@guolinke yes, I forgot adding it. btw, there are some files for when proto=on, some files for when proto=off, they are controlled by cmake, should I add them all to windows/LightGBM.vcxproj?

**Comment by guolinke** [comment]
@wxchan it seems VS cannot support proto, so we can just disable it

**Pull request merged**